observation_portal/proposals/viewsets.py

```diff
@@ -34,3 +78,130 @@ class SemesterViewSet(viewsets.ReadOnlyModelViewSet):
     ordering = ('-start',)
     queryset = Semester.objects.all()
 
+    @action(detail=True, methods=['get'])
+    def proposals(self, request, pk=None):
+        semester = self.get_object()
+        proposals = semester.proposals.filter(
+            active=True, non_science=False
+        ).prefetch_related(
+            'sca', 'membership_set', 'membership_set__user', 'membership_set__user__profile',
+            'semester_set', 'timeallocation_set'
+        ).distinct().order_by('sca__name')
+        results = []
+        for proposal in proposals:
+            results.append({
+                'id': proposal.id,
+                'title': proposal.title,
+                'abstract': proposal.abstract,
+                'allocation': proposal.allocation(semester=semester),
+                'pis': [
+                    {
+                        'first_name': mem.user.first_name,
+                        'last_name': mem.user.last_name,
+                        'institution': mem.user.profile.institution
+                    } for mem in proposal.membership_set.all() if mem.role == Membership.PI
+                ],
+                'sca_id': proposal.sca.id,
+                'sca_name': proposal.sca.name,
+                'semesters': proposal.semester_set.distinct().values_list('id', flat=True)
+            })
+        return Response(results)
+
+    @action(detail=True, methods=['get'], permission_classes=(IsAdminUser,))
+    def timeallocations(self, request, pk=None):
+        timeallocations = self.get_object().timeallocation_set.prefetch_related(
+            'proposal', 'proposal__membership_set', 'proposal__membership_set__user'
+        ).distinct()
+        results = []
+        for timeallocation in timeallocations:
+            memberships = timeallocation.proposal.membership_set
+            timeallocation_dict = timeallocation.as_dict(exclude=['proposal', 'semester'])
+            timeallocation_dict['proposal'] = {
+                'notes': timeallocation.proposal.notes,
+                'id': timeallocation.proposal.id,
+                'tac_priority': timeallocation.proposal.tac_priority,
+                'num_users': memberships.count(),
+                'pis': [
+                    {
+                        'first_name': mem.user.first_name,
+                        'last_name': mem.user.last_name
+                    } for mem in memberships.all() if mem.role == Membership.PI
```

Comment:
you could do memberships.filter(role==Membership.PI) here instead of all and the conditional

Reply:
I had started with that, but for some reason doing it this way ended up actually being quite a bit faster. But you're right, that is a more natural way of doing it. I'll get silk set up and do the speed test again - maybe there was something that I had forgotten to prefetch or something.

Reply:
interesting... maybe it applies the filter() to everything before pairing it down to the membership_set only?

Reply:
I set up silk, and indeed many more queries are executing when I do `memberships.filter(role==Membership.PI)`. The reason is that prefetch_related does `.all()` of whatever is being prefetched, and since `.filter(...)` is a different query from `.all()`, the prefetched results don't get used and a new query is kicked off for each object instead. I'm going to leave as is.